src/main/java/org/testng/annotations/SkipCloning.java

```diff
@@ -0,0 +1,16 @@
+package org.testng.annotations;
```

Comment:
@juherr - Here's my rationale behind leaving this exposed ( I also would like to evangelize about this lesser known behavior of TestNG )

Currently when it comes to parameter injection and associating that with the `TestResult` object, we resort to `cloning()` if the user's parameter object class implements the `Cloneable` interface. So there could be a chance wherein the parameters that the user is altering in their tests is perhaps not reflected in their listeners or in their reports because we cloned them.

I think we should enrich our documentation to first call out this behavior explicitly ( a most common use case is a data driven test that is run via a `@DataProvider` annotation) and via this newly introduced annotation, still let the user control whether they want or do not want their test objects to be cloned by TestNG.

Reply:
I disagree because you want to make public something which wasn't asked.
Then, the clone solution was just a quick workaround for a design issue when the same object is used in reports. But the clone solution doesn't cover all the need.

That's why I have some doubts to publish a workaround of a workaround before be really sure how we need to fix the root issue.

Reply:
> I disagree because you want to make public something which wasn't asked.

The changeset that introduced setting cloned variants of parameters was to fix https://github.com/cbeust/testng/issues/447 It was done by you via commit https://github.com/cbeust/testng/commit/78001fd0ed81970784b9d17b887998064fff0c03

Via this commit we ended up implicitly cloning a parameter (if it was clone able) but never made this information public. IMO we should have done this, because an end-user should know if the real parameter that is being passed to the test method is being used or if its a cloned one. 

We haven't heard of anyone mention about this problem only because usually our users dont implement the `Cloneable` interface to the parameter that they are passing to a data provider.

The moment someone does that, this implicit behavior will become evident. 

So now it looks like the root cause was us introducing the Cloneable parameter.  
What if the user provided parameter was clone able but the user didn't want it to be clone able ? How would they be able to support that. With collections it may not be possible because a user cannot add that annotation to a collection class (I think this also should be possible if we just enrich `@SkipCloning` annotation such that it can be added to a parameter just as how a user adds `@Optional`), but at least with a user's test project defined objects it should be something that they should be able to do.

This was my thinking.

Reply:
Yes, I understand your point of view. But mine is the #447 was just a not complete workaround of the root issue.
I agree to release another workaround to fix the first workaround but I'm not feeling it good to make it public because we will have to manage deprecation.
Instead, I prefer to make it private and to find a good solution for the root issue.
Furthermore, I'm not sure my old commit will still work on with Java9 modules due to` clone.setAccessible(true)`.

For the current PR:
* private: OK and waiting for a better fix for #447
* public: not yet ok :)

Reply:
ping @juherr

Reply:
Without giving the control of deciding which parameter to be cloned and which not to, how do you propose the problem will be fixed ? Can you please help elaborate that ? 

@cbeust what are you thoughts on this ?